### ISSUE TITLE
SGE issues

### ISSUE THREAD

**Issue by tatarsky** [frazer-lab/cluster#211]
sge_qmaster exploded in memory consumption for some reason a few minutes ago.

I am looking at it but SGE is down until I can figure it out.

If any usual items were being done just advise.

**Comment by tatarsky** [comment]
Did somebody submit a very large number of jobs?

```
08/11/2017 12:36:51|  main|fl-hn1|I|read job database with 15003 entries in 19 seconds
```

**Comment by billgreenwald** [comment]
Yes Joaquin submitted a ton

**Comment by tatarsky** [comment]
Whats "a ton" ?

**Comment by tatarsky** [comment]
I assume ~15K. Thats just not a good thing. Any reason they are all individual jobs and not an array? Attempting to recover from it.

**Comment by billgreenwald** [comment]
I can't tag him from mobile but he could answer that.  He did ask what the max number to submit was and we didn't know. Do you have a general rule of thumb?

**Comment by tatarsky** [comment]
It varies. Whatever he's submitting is having extreme impact. I can diagnose once I recover. I'm setting a cap for now down to 5000. 

I guess I normally see array jobs as a far far more efficient way to submit than what I'm seeing in the spool files. When he gets online he can chat here and we can analyze it a bit. 

But basically I see thousands of the same looking script, which I often then assume would be running over a series of changed items. And array jobs are far less impact than single jobs.

**Comment by tatarsky** [comment]
Jobs will likely have to be deleted to regain interactive response. The job count is resulting in an sge_qmaster consuming basically 50% of the head node ram. Thats not a great situation. If you want to tough it out, can let it ride but basically this effort should be reviewed with me for better submit methods.

**Comment by tatarsky** [comment]
To be clear: I would delete pending joreyna jobs.

**Comment by tatarsky** [comment]
qstat -u joreyna|wc -l
16122

**Comment by joreynajr** [comment]
Hi Paul,

You can kill them.

**Comment by tatarsky** [comment]
Will try but will take awhile.

**Comment by joreynajr** [comment]
Hiroko recommended I use array submission. It's might fault. I haven't used array submission much but I'll look into it so this doesn't happen again.

**Comment by tatarsky** [comment]
Array jobs are FAR more efficient and produce less per item memory footprint. I am always happy to review submits that could push the limits. When I say "it varies" there are some job options which can produce even more memory impact (I believe dependencies in particular).

**Comment by tatarsky** [comment]
Started the qdel.

**Comment by tatarsky** [comment]
Judging from the ganglia SGE graphs and the memory graphs I'm going to say that ~8000 non-array jobs of this is pushing our resources too far. 

If array jobs do NOT work for you, I often recommend a small script that feeds jobs in 1000 or so batches and makes sure they are exiting before starting the next batch. 

Bearing in mind, array jobs have that concept built in ;)

**Comment by tatarsky** [comment]
Here's a basic array job tutorial grabbed from Duke.

https://wiki.duke.edu/display/SCSC/SGE+Array+Jobs

The gist is that when using array jobs via something like `qsub -t 1-10000 mylotsofstuff.sh` the environment variable SGE_TASK_ID will be set for each array job member to the sequence 1->10000.

What you do with that is up to you. Many people write a small handler which takes that increment variable and changes the arguments for that subtask. 

I can often help adapt a stream of single jobs to an array job if you just explain what method you use to make the stream of single jobs. Just drop me a note or we can talk about it here. Up to you.

**Comment by tatarsky** [comment]
sge_qmaster ran out of memory. I may need to move the spool aside to fix this. Which affects running jobs. Trying again.

**Comment by tatarsky** [comment]
So cleared the jobs mostly just for @joreynajr and looking at the submit scripts. 

So far what I see is basically this is the same issue as back in Git #170  and the culprit again is the use of `-hold_jid` and its memory consumption which is large for reasons only an SGE developer could explain probably but I know it really cranks up the per job memory footprint in the qmaster. We hit the same issue I believe with around 11K jobs. 

I believe these also might be the same scripts. 

But in that Git thread the statement was made: "I made a change in my code try not to use -hold_jid for submitting many jobs." That might be worth understanding why so many dependencies on other jobs are being made. 

I'll look as I can but I have a dinner appointment here shortly.

Basically "beware of hold_jid" when submitting many many jobs is our statement for now.

**Comment by tatarsky** [comment]
I will review also the "max jobs" options in SGE in a way that doesn't impact array jobs. But I've removed the cap for now as I have to recall the exact settings. So please be careful as I head into my weekend particular with -hold_jid. That alters dramatically the impact of submitting lots of jobs.

**Comment by tatarsky** [comment]
This has been resolved by not issuing so many jobs with holds based on other jobs. If holds are truly needed a similar concept exists for array jobs (`-hold_job_ad`) which requires arrays have the same size but is still interesting. I do not know if it has the same qmaster impact so if you wish to try that method please start with small array loops. man qsub has some details in its section on that flag.

Alternately, a pipeline that stages its jobs with test routines between them is often what I see to do the same "concept" as a job dependency. Aka only start the second batch when a test script has confirmed the output from the first is done and also _correct_.
